peer/internal/uprng/uprng.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2024 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package uprng
+
+import (
+	cryptorand "crypto/rand"
+	"encoding/binary"
+	"io"
+	"math/bits"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/chacha20"
+)
+
+// Reader implements a cryptographically secure userspace PRNG that is
+// periodically reseeded with entropy obtained from crypto/rand.
+// Reader is safe for concurrent access.
+var Reader io.Reader
+
+// Read fills b with random bytes obtained from the userspace PRNG.
+//
+// When possible, prefer Read over using Reader to avoid escape analysis
+// unnecessarily allocating b's backing array on the heap.
+func Read(b []byte) {
+	defaultReader.Read(b)
+}
+
+var defaultReader = newPRNG()
+
+func init() {
+	Reader = defaultReader
+}
+
+const (
+	maxCipherRead     = 4 * 1024 * 1024 // 4 MiB
+	maxCipherDuration = 20 * time.Second
+)
+
+// nonce implements a 12-byte little endian counter suitable for use as an
+// incrementing ChaCha20 nonce.
+type nonce struct {
```

Comment:
Super minor thing, but based on my `secp256k1` work, I'm about 99.9% positive this would be faster to just grab the limbs from the bytes like so because the compiler is able to optimize it better:

```Go
type nonce [chacha20.NonceSize]byte
func (n *nonce) inc() {
	n0 := binary.LittleEndian.Uint32(n[0:4])
	n1 := binary.LittleEndian.Uint32(n[4:8])
	n2 := binary.LittleEndian.Uint32(n[8:12])

	var carry uint32
	n0, carry = bits.Add32(n0, 1, carry)
	n1, carry = bits.Add32(n1, 0, carry)
	n2, carry = bits.Add32(n2, 0, carry)
	n0, _ = bits.Add32(n0, 0, carry)

	binary.LittleEndian.PutUint32(n[0:4], n0)
	binary.LittleEndian.PutUint32(n[4:8], n1)
	binary.LittleEndian.PutUint32(n[8:12], n2)
}
```

Reply:
Just tested it:

```
BenchmarkNonceIncOrig       553819740                2.171 ns/op           0 B/op          0 allocs/op
BenchmarkNonceIncNew        683211388                1.718 ns/op           0 B/op          0 allocs/op
```

Reply:
Benchstat with `-count=100` to make sure it's not a fluke:

```
NonceIncOrig  2.18ns ± 2%
NonceIncNew   1.72ns ± 1%
```